### ISSUE TITLE
Unnecessary return value of `BaseStorage.set_trial_param`.

### ISSUE THREAD

**Issue by ytsmiling** [optuna/optuna#1308]
## TL;DR
This issue requests to deprecate the return value of `BaseStorage.set_trial_param` and change the behavior to "always overwrite existing values".

## General background
[`BaseStorage.set_trial_param`](https://github.com/optuna/optuna/blob/24300dc57661bff7961499a2d52534fcd41ad258/optuna/storages/base.py#L120) method returns True and set a new parameter when the specified parameter-name does not already have a value. Otherwise, it returns False and does not overwrite the parameter with the new value.
This behavior was necessary to prevent race-conditioning. However, the current optuna doesn't allow users to directly manipulate the storage and optuna internals do not modify parameters from multiple workers/threads, and thus, the functionality is no more required.

## Proposal and motivation
This issue requests to deprecate the return value of `BaseStorage.set_trial_param` and change its behavior to "always overwrite existing values".
Instead of using the return value, we'd better check whether the parameter already exists first by ``param_name in storage.get_trial_params(trial_id)`` and then perform sampling operation in [this method in trial](https://github.com/optuna/optuna/blob/24300dc57661bff7961499a2d52534fcd41ad258/optuna/trial.py#L981)
This change has two benefits.
1. We can remove unnecessary computation for re-sampling when the parameter value already exists.
2. The storage method becomes more consistent with other methods such as `set_trial_user_attrs`, which always overwrite existing values. It also simplifies some storage implemetations.
